datalake/coordinator: group duplicate ensure_table calls

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
In certain scenarios with a large number of partitions I would see that
sometimes translation would get stuck, with RPCs to the coordinator all
timing out. I suspect that this is because of the large volume of
ensure_table_schema requests that could come in, particularly right
after a restart, when all translators are starting up together.

To mitigate this, this commit adds tracking of on-going requests to
ensure_table_schema and groups incoming requests in with on-going ones,
if they exist.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
datalake/coordinator: group duplicate ensure_table calls

In certain scenarios with a large number of partitions I would see that
sometimes translation would get stuck, with RPCs to the coordinator all
timing out. I suspect that this is because of the large volume of
ensure_table_schema requests that could come in, particularly right
after a restart, when all translators are starting up together.

To mitigate this, this commit adds tracking of on-going requests to
ensure_table_schema and groups incoming requests in with on-going ones,
if they exist.

andrwng committed:
iceberg/rest_client: don't throw on closed gate

Avoids a closed gate exception from the catalog client. Since we are
using a result type (expected<>), it's easier to reason about the code
when we don't have to think about exceptions too _and_ error types.

WillemKauf commented on src/v/datalake/coordinator/coordinator.h:
```diff
@@ -88,6 +90,46 @@ class coordinator {
+              std::move(h), k.topic, k.topic_revision, k.record_comps);
+        }
+    };
+    struct table_key_hash {
```
`table_key_hash` and `table_key_eq` are currently unused - are these meant to be used in the `using` typedef for `ensure_table_map_t`?

andrwng replied:
Unneeded copy pasta, nice catch!

WillemKauf reviewed (dismissed):
`notify_waiters_and_erase()` and `maybe_add_waiter()` feel like nice generics to have for use with any request type- do you think we will want to extend this pattern to future request types beyond `ensure_table_result_t`? Or is this the only "problematic" request type we are constantly hammering the catalog with?

WillemKauf reviewed (dismissed):
Would it be beneficial/easy to add some sort of fixture test for `maybe_add_waiter()` and `notify_waiters_and_erase()`?

Not a blocker, just wondering. I assume you've already seen empirical results which show the change working/reducing stress on the coordinator

andrwng commented:
> `notify_waiters_and_erase()` and `maybe_add_waiter()` feel like nice generics to have for use with any request type- do you think we will want to extend this pattern to future request types beyond `ensure_table_result_t`? Or is this the only "problematic" request type we are constantly hammering the catalog with?

Yea something generic makes sense. We have a utils/waiter_queue that's already somewhat similar. As for other RPCs, I think we're good for now, the ensure_table RPCs are the only ones that end up immediately hitting the catalog.

There's probably room in the schema manager on the translators to have similar logic when loading the table from the catalog. Will consider that for a follow up.

andrwng commented:
> Would it be beneficial/easy to add some sort of fixture test for `maybe_add_waiter()` and `notify_waiters_and_erase()`?
> 
> Not a blocker, just wondering. I assume you've already seen empirical results which show the change working/reducing stress on the coordinator

Yea, makes sense. Will see about this on Monday.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--591dbfd3-7bec-4d23-8bb4-b135725a42d2-->
<!--TEST RESULTS-->
<!--35a6c8e8-68ec-41f9-93f3-099a87bd6792-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63931">build#63931</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_avro_schema.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7db8-4c21-94b1-1ba4443eea6c | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e34b-39e0-4dce-be7a-360327cbddb3 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7db6-41af-86d8-0e73c88b3943 | FLAKY | 1/3 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_select_fails.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7dba-40ab-8a93-16a83691ae20 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_select_fails.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7db8-4c21-94b1-1ba4443eea6c | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_illegal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=illegalpromotionint-.string.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e34b-39e0-4dce-be7a-360327cbddb3 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_illegal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=illegalpromotionint-.string.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7dba-40ab-8a93-16a83691ae20 | FLAKY | 1/3 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_illegal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=illegalpromotionint-.string.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e34b-39de-453d-82d4-3a16b115c3cb | FLAKY | 1/3 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=add_column.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e34b-39dd-4e99-a130-34c3374e28cd | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=drop_column.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7db6-41af-86d8-0e73c88b3943 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=promote_column.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e34b-39dd-4e99-a130-34c3374e28cd | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=promote_column.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e34b-39e0-4dce-be7a-360327cbddb3 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=promote_column.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7db6-41af-86d8-0e73c88b3943 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=promote_column.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e34b-39df-4052-8964-8c873b741e7c | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=drop_column.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e34b-39e0-4dce-be7a-360327cbddb3 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=drop_column.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7db6-41af-86d8-0e73c88b3943 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=promote_column.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7db8-4c21-94b1-1ba4443eea6c | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_old_schema_writer.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=add_column.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7db8-4c21-94b1-1ba4443eea6c | FLAKY | 1/3 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_old_schema_writer.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=drop_column.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e34b-39df-4052-8964-8c873b741e7c | FLAKY | 1/3 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_old_schema_writer.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=reorder_columns.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e34b-39de-453d-82d4-3a16b115c3cb | FLAKY | 1/3 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_old_schema_writer.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=reorder_columns.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7db8-4c21-94b1-1ba4443eea6c | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_reorder_columns.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e34b-39df-4052-8964-8c873b741e7c | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_reorder_columns.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7db8-4c21-94b1-1ba4443eea6c | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_reorder_columns.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7db6-41af-86d8-0e73c88b3943 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_reorder_columns.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63931#0195e350-7dba-40ab-8a93-16a83691ae20 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63944">build#63944</a></summary>


| test_id                                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves.cleanup_policy=compact.delete | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63944#0195e5f5-d733-4213-aac1-596e1561b2ce | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63944#0195e5f5-d736-4544-9364-dbaa37d28656 | FLAKY         | 1/2      |
</details>
<!--/body-->

WillemKauf commented:
Lot of flakes in CI, probably a bug somewhere?

dotnwat commented on src/v/iceberg/rest_client/catalog_client.cc:
```diff
@@ -172,6 +172,13 @@ catalog_client::maybe_configure(retry_chain_node& rtc) {
     co_return std::monostate{};
 }
 
+expected<ss::gate::holder> catalog_client::maybe_gate() {
+    if (_gate.is_closed()) {
+        return tl::unexpected(aborted_error("catalog_client::_gate closed"));
+    }
+    return _gate.hold();
+}
```
neat

andrwng commented:
Moved to draft while debugging flakiness

andrwng committed:
iceberg/rest_client: fix retry_chain_node usage

The retry_chain_node in catalog_client requests was being reused across
all requests made within a single top-level operation. This meant that
for a single e.g. create_table, each time we got an OAuth token, or got
the catalog config, or performed the create table request, it would
count towards our retry count, and cause subsequent calls to miscompute
whether the retry node is allowed to retry.

Fixes this by splitting off a child node for each request instead. This
makes each request have its own retry count while sharing the same
deadline and abort source.

WillemKauf commented on src/v/iceberg/rest_catalog.cc:
```diff
@@ -98,15 +98,17 @@ ss::future<checked<table_metadata, catalog::errc>> rest_catalog::create_table(
   const partition_spec& spec) {
-    auto rtc = create_rtc();
+    auto parent_rtc = create_rtc();
+    auto table_rtc = retry_chain_node(&parent_rtc);
```
nit: I wonder if these child `rtc`s could be more tightly scoped if we had a function like `retry_chain_node::make_child()` which would basically return the result of `retry_chain_node::retry_chain_node(retry_chain_node* parent)`- then the change would only be 

```cpp
auto res = co_await client_->create_table(
      t_id.ns, std::move(request), rtc.make_child());
```

andrwng replied:
Yea i think that makes sense. Seems like a missing part of the API that makes it easy for these kinds of bugs to happen, cc @lazin

WillemKauf replied:
I guess part of the reason this doesn't already exist is the [existing laundry list of `retry_chain_node` constructors which take in a `retry_chain_node* parent`](https://github.com/redpanda-data/redpanda/blob/438336f27aeabac766990366fc8693271eb008ea/src/v/utils/retry_chain_node.h#L234-L265) which `make_child()` (and it's potential overloads) could decide to support :melting_face:

WillemKauf reviewed (approved):
CI looks good :+1:

vbotbuildovich commented:
/backport v25.1.x
